matrix/Quaternion.hpp

```diff
@@ -254,23 +266,40 @@ class Quaternion : public Vector<Type, 4>
     }
 
     /**
+    * Apply an angular velocity vector to this quaternion and normalize result
+    */
+    void applyRates(const Matrix31 &w, float dt) {
+        Quaternion<Type> &p = *this;
+        p = p + derivative(w) * dt;
+
+        // automatically normalize if error is larger than some constant (normalize before use for best result)
+        // NOTE: if normalization is done always then it creates horrible precision because of floating point roundoffs
+        if(fabs(p.dot(p) - 1.0f) > 0.005) {
+            p.normalize();
+        }
+    }
+
+    /**
      * Invert quaternion
      */
-    void invert()
-    {
+    void invert() {
         Quaternion &q = *this;
         q(1) *= -1;
         q(2) *= -1;
         q(3) *= -1;
     }
 
```

Comment:
I think I had to cast some intermediate value explicitly to Quaternion because compiler was unable to convert Vector4 to quaternion automatically in that context.